File: solutions/typescript/list-ops/1/list-ops.ts
class Node<T> {
  data: T;
  next?: Node<T>;

  constructor(data: T, next?: Node<T>) {
    this.data = data;
    this.next = next;
  }
}

type Predicate<T> = (el: T) => boolean;

export class List<T> {
  head?: Node<T>;
  tail?: Node<T>;

  public static create<T>(...values: T[]): List<T> {
    const list = new List<T>();

    for (const value of values) {
      list._push(value);
    }

    return list;
  }

  private _push(el: T) {
    const node = new Node(el);
    if (!this.head) this.head = node;
    if (this.tail) this.tail.next = node;

    this.tail = node;
  }

  private _insert(el: T) {
    const node = new Node(el);
    if (!this.tail) this.tail = node;
    if (this.head) node.next = this.head;
    this.head = node;
  }

  append(list: List<T>) {
    if (!list.head) return this;

    if (!this.head) {
      this.head = list.head;
      this.tail = list.tail;
      return this;
    }

    if (this.tail) this.tail.next = list.head;
    this.tail = list.tail;

    return this;
  }

  concat(list: List<List<T>>) {
    if (!list.head) return this;
    list.forEach((el) => this.append(el));
    return this;
  }

  filter(fn: Predicate<T>): List<T> {
    const list = new List<T>();

    this.forEach((el) => {
      if (fn(el)) list._push(el);
    });

    return list;
  }

  length(): number {
    let c = 0;
    this.forEach(() => c++);
    return c;
  }

  map(fn: (el: T) => T): List<T> {
    const list = new List<T>();
    this.forEach((el) => list._push(fn(el)));
    return list;
  }

  foldl(fn: (acc: any, el: T) => any, initialValue: any) {
    let acc: any = initialValue;
    this.forEach((el) => (acc = fn(acc, el)));
    return acc;
  }

  foldr(fn: (acc: any, el: T) => any, initialValue: any) {
    const list = this.reverse();
    return list.foldl(fn, initialValue);
  }

  reverse(): List<T> {
    const list = new List<T>();
    this.forEach((el) => list._insert(el));
    return list;
  }

  forEach(fn: (el: T) => any) {
    let node = this.head;

    while (node) {
      fn(node.data);
      node = node.next;
    }
  }
}

File: solutions/elixir/list-ops/1/lib/list_ops.ex
defmodule ListOps do
  # Please don't use any external modules (especially List or Enum) in your
  # implementation. The point of this exercise is to create these basic
  # functions yourself. You may use basic Kernel functions (like `Kernel.+/2`
  # for adding numbers), but please do not use Kernel functions for Lists like
  # `++`, `--`, `hd`, `tl`, `in`, and `length`.

  @spec count(list) :: non_neg_integer
  def count(l) do
    foldl(l, 0, fn _, acc -> acc + 1 end)
  end

  @spec reverse(list) :: list
  def reverse(l) do
    foldl(l, [], fn item, acc -> [item | acc] end)
  end

  @spec map(list, (any -> any)) :: list
  def map(l, f) do
    foldr(l, [], fn item, acc -> [f.(item) | acc] end)
  end

  @spec filter(list, (any -> as_boolean(term))) :: list
  def filter(l, f) do
    foldr(l, [], fn item, acc -> if f.(item), do: [item | acc], else: acc end)
  end

  @type acc :: any
  @spec foldl(list, acc, (any, acc -> acc)) :: acc
  def foldl([], acc, _), do: acc

  def foldl([h | t], acc, f), do: foldl(t, f.(h, acc), f)

  @spec foldr(list, acc, (any, acc -> acc)) :: acc
  def foldr(l, acc, f) do
    l |> reverse() |> foldl(acc, f)
  end

  @spec append(list, list) :: list
  def append(a, b) do
    foldr(a, b, fn item, acc -> [item | acc] end)
  end

  @spec concat([[any]]) :: [any]
  def concat(ll) do
    foldr(ll, [], &append/2)
  end
end
